keps/sig-api-machinery/3037-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,590 @@
+
+Client-go is the library used by all the cluster components to communicate with the API servers, any change or regression there can have a great impact on the project.
+
+The implementation of this feature should not modify any of current behaviors and be completely pluggable, so users can opt-in safely and the project can be sure there will not create any regression.
+
+### Risks and Mitigations
+
+The feature will be opt-in, users can disable it and keep having the same behavior.
+
+To avoid possible security issues, this feature will require HTTP2 and HTTPS in order to work.
+
+The alternative services IPs exposed are the same published in the Endpoints for the `kubernetes.default` services, the Alt-Svc headers will not be present on requests
+that are not authenticated.
+
+## Design Details
+
+There are two problems we have to solve:
+
+- get a list of available control-plane nodes
+- client capability to be able to connect to multiple control-planes
+
+### List of available control-plane nodes
+
+This problem can be solved in two different ways:
+
+1. Configuring on the client the list of available control-plane nodes
+
+This can be done extending the client-go configuration with an additional field that allows users to include a list of alternative control-plane nodes that the client can use.
+
+`k8s.io/client-go/rest/config.go`
+```go
+type Config struct {
+	// Host must be a host string, a host:port pair, or a URL to the base of the apiserver.
+	// If a URL is given then the (optional) Path of that URL represents a prefix that must
+	// be appended to all request URIs used to access the apiserver. This allows a frontend
+	// proxy to easily relocate all of the apiserver endpoints.
+	Host string
+  // AlternativeHosts must be a comma separated list if hosts, host:port pairs or URLs to the base of
+  // different apiservers. The client can use any of them to access the apiserver.
+	AlternativeHosts string
+```
+
+2. RFC7838 HTTP Alternative Services
+
+This specification defines a new concept in HTTP, "Alternative Services", that allows an origin server to nominate additional means of interacting with it on the network.
+
+All conformant clusters are required to publish a list of endpoints with the apiserver addresses.In the default implementation, this list of endpoints is generated by a reconcile loop that guarantees that only ready apiserver addresses are present, see https://github.com/kubernetes/kubernetes/tree/master/pkg/controlplane/reconcilers.
+
+The proposal is for API servers to have an option to enable the generation of the Alternative Services headers based on the list of Endpoints created for the `kubernetes.default` service, this requires that the users of this feature belong to the cluster network. TODO define a way of filtering the headers in the answer, per example, by source IP, certificate name, ... 
```

Comment:
is the same problem that with endpoints, services, load balancers, dns, when you publish an IP you are not making a contract that guarantees connectivity, is information .... though, we can limit to whom we want to provide this information

Reply:
I revisited the RFC to confirm that clients do not need to use the alternative service and , in case of using it, they can fall back https://datatracker.ietf.org/doc/html/rfc7838#section-2.4

The user should opt-in for this feature, and in case it does opt but the client is not able to reach those ips, the roundtripper is a noop, it blacklist the ips and use the origina one ... I tested that thoroughly